server/src/test/java/org/elasticsearch/action/admin/cluster/snapshots/get/SnapshotNamePredicateTests.java

```diff
@@ -0,0 +1,137 @@
+    }
+
+    public void testIncludeWildcardExcludeName() {
+        final var requestName = randomIdentifier();
+        final var predicates = createPredicate(randomBoolean(), "include-*", "-include-" + requestName);
+        assertTrue(predicates.test("include-" + randomValueOtherThan(requestName, ESTestCase::randomIdentifier), randomBoolean()));
+        assertFalse(predicates.test("include-" + requestName, randomBoolean()));
+        assertThat(predicates.requiredNames(), empty());
+    }
+
+    public void testIncludeWildcardExcludeWildcard() {
+        final var predicates = createPredicate(randomBoolean(), "include-*", "-include-exclude-*");
+        assertTrue(predicates.test("include-" + randomIdentifier(), randomBoolean()));
+        assertFalse(predicates.test("exclude-" + randomIdentifier(), randomBoolean()));
+        assertFalse(predicates.test("include-exclude-" + randomIdentifier(), randomBoolean()));
+        assertThat(predicates.requiredNames(), empty());
+    }
+
+    public void testIncludeCurrentExcludeWildcard() {
+        final var predicates = createPredicate(randomBoolean(), currentSnapshotsSelector(), "-exclude-*");
+        assertTrue(predicates.test(randomIdentifier(), true));
+        assertFalse(predicates.test("exclude-" + randomIdentifier(), randomBoolean()));
+        assertFalse(predicates.test(randomIdentifier(), false));
+        assertThat(predicates.requiredNames(), empty());
+    }
+
+    public void testIncludeCurrentAndWildcardExcludeName() {
+        final var requestName = randomIdentifier();
+        final var predicates = createPredicate(randomBoolean(), currentSnapshotsSelector(), "include-*", "-include-" + requestName);
+        assertTrue(predicates.test(randomIdentifier(), true));
+        assertTrue(predicates.test("include-" + randomValueOtherThan(requestName, ESTestCase::randomIdentifier), false));
+        assertFalse(predicates.test("include-" + requestName, randomBoolean()));
+        assertThat(predicates.requiredNames(), empty());
+    }
+
+    public void testInitialExclude() {
+        // NB current behaviour, but could be considered a bug?
+        final var requestName = "-" + randomIdentifier();
+        final var predicates = createPredicate(false, requestName);
+        assertTrue(predicates.test(requestName, randomBoolean()));
+        assertThat(predicates.requiredNames(), equalTo(Set.of(requestName)));
+    }
+
+    public void testHyphen() {
+        // NB current behaviour, but could be considered a bug?
+        final var predicates = createPredicate(false, "include-*", "-");
+        assertTrue(predicates.test("include-" + randomIdentifier(), randomBoolean()));
+        assertTrue(predicates.test("-", randomBoolean()));
+        assertThat(predicates.requiredNames(), equalTo(Set.of("-")));
+    }
+
+    public void testAllWithExclude() {
+        // NB current behaviour, but could be considered a bug?
```

Comment:
Can we forbid snapshot names to start with `-`?

Reply:
If anyone is actually naming snapshots with a `-` prefix then they're going to be having a pretty bad time of it with APIs like this, but I think it's not forbidden today (see `org.elasticsearch.snapshots.SnapshotsService#validate`) and to start forbidding it would be a breaking change.